### PR TITLE
Properly support batched/non-batched with vllm/llama.cpp 

### PR DESCRIPTION
**Pull request by njhill**
This is based on @npalaska's PR #58.

With these changes we will auto-detect whether the server supports batched inputs and if not will send them sequentially.

### CODE REVIEW

File: src/instructlab/sdg/llmblock.py [hunk 1]
```diff
@@ -84,27 +90,40 @@ def _parse(self, generated_string) -> dict:
...
+            )
+            return [choice.text.strip() for choice in response.choices]
+
+        n = gen_kwargs.get("n", 1)
+        results = []
+        for prompt in prompts:
+            for _ in range(n):
+                response = self.client.completions.create(
+                    prompt=prompt, **generate_args
+                )
```
Comment (npalaska) [inline src/instructlab/sdg/llmblock.py:115]: We could rewrite this as 
```
responses = [
    self.client.completions.create(prompt=prompt, **generate_args)
    for prompt in prompts
    for _ in range(n)
]
```
wdyt?

File: src/instructlab/sdg/llmblock.py [hunk 2]
```diff
@@ -167,29 +194,33 @@ def __init__(
...
+
+        return self.prompt_template.format(**sample).strip()
 
     def validate(self, prompt_template: str, input_dict: Dict[str, Any]) -> bool:
         if isinstance(prompt_template, dict):
             prompt_template = prompt_template[input_dict[self.selector_column_name]]
         return super()._validate(prompt_template, input_dict)
+
+
+def server_supports_batched(client, model_id: str) -> bool:
```
[the remark this answers is not in the document]
Comment (njhill) [reply src/instructlab/sdg/llmblock.py:210]: I think batched is better.. since it's referring to the inputs. Even without batched inputs it might do batching internally.

File: src/instructlab/sdg/llmblock.py (hunk 1 above)
Comment (njhill) [reply src/instructlab/sdg/llmblock.py:115]: yes but then we would require an additional loop anyhow

File: src/instructlab/sdg/llmblock.py [hunk 3]
```diff
@@ -113,24 +132,32 @@ def generate(self, samples, **gen_kwargs) -> Dataset:
...
+            extended_samples.extend([item] * num_parallel_samples)
+
+        print(f"num outputs is {len(outputs)}")
         new_data = []
-        for sample, output in zip(samples, outputs):
+        for sample, output in zip(extended_samples, outputs):
             parsed_outputs = self._parse(output)
-            # pylint: disable=consider-using-generator
+
             max_length = max([len(value) for value in parsed_outputs.values()])
```
[the remark this answers is not in the document]
Comment (njhill) [reply src/instructlab/sdg/llmblock.py:156]: This logic still needs to be cleaned up anyhow I think, it's not doing what it was intended to

**Comment by njhill** [comment]
Thanks @npalaska I addressed most of those comments.

File: src/instructlab/sdg/llmblock.py (hunk 1 above)
Comment (npalaska) [reply src/instructlab/sdg/llmblock.py:115]: Yeah

File: src/instructlab/sdg/llmblock.py [hunk 4]
```diff
@@ -45,8 +46,13 @@ def __init__(
             "model": self.model,
             "temperature": 0,
             "max_tokens": 12000,
+            #"seed": 12345,  TBD
         }
 
+        # Whether the LLM server supports a list of input prompts
+        # and supports the n parameter to generate n outputs per input
+        self.server_supports_batched = server_supports_batched(client, model_id)
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:54]: The `FlowParams` in #64 would give us a place to do this once rather than for every `LLMBlock`, but that can be fixed up later

File: src/instructlab/sdg/llmblock.py [hunk 5]
```diff
@@ -45,8 +46,13 @@ def __init__(
             "model": self.model,
             "temperature": 0,
             "max_tokens": 12000,
+            #"seed": 12345,  TBD
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:49]: Delete? Or add an explanation to the comment

File: src/instructlab/sdg/llmblock.py [hunk 6]
```diff
@@ -84,27 +90,40 @@ def _parse(self, generated_string) -> dict:
...
-        return [choice.text.strip() for choice in response.choices]
+        generate_args = {**self.defaults, **gen_kwargs}
+
+        if self.server_supports_batched:
+            response = self.client.completions.create(
+                prompt=prompts, **generate_args
+            )
+            return [choice.text.strip() for choice in response.choices]
+
+        n = gen_kwargs.get("n", 1)
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:109]: I would have imagined doing this in reverse - including "num_instructions_to_generate" in the block config and adding 'n' to gen_kwargs if batching was supported. No biggie though

File: src/instructlab/sdg/llmblock.py [hunk 7]
```diff
@@ -113,21 +132,30 @@ def generate(self, samples, **gen_kwargs) -> Dataset:
         # validate each sample
         for sample in samples:
             if not self._validate(self.prompt_template, sample):
-                return None
+                logger.warning("Sample failed validation")  #TODO add details
+                #TODO remove sample from samples
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:136]: Hmm. Should this be in a separate PR. If in this PR, the TODO should be resolved?

File: src/instructlab/sdg/llmblock.py [hunk 8]
```diff
@@ -113,21 +132,30 @@ def generate(self, samples, **gen_kwargs) -> Dataset:
...
-        if batched:
-            outputs = self._generate(samples, **gen_kwargs)
-        else:
-            outputs = [self._generate([sample], **gen_kwargs)[0] for sample in samples]
-        logger.debug("Generated outputs: {}".format(outputs))
+
+        outputs = self._generate(samples, **gen_kwargs)
+        logger.debug("Generated outputs: %s", outputs)
+
+        num_parallel_samples = gen_kwargs.get("n", 1)
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:146]: Hmm, and here's a reason to make num_parallel_samples part of the block config ... and add 'n' to gen_kwargs based on that

File: src/instructlab/sdg/llmblock.py [hunk 9]
```diff
@@ -167,29 +195,33 @@ def __init__(
...
+        # Make a test call to the server to determine whether it supports
+        # multiple input prompts per request and also the n parameter
+        response = client.completions.create(
+            model=model_id, prompt=["test1", "test2"], max_tokens=1, n=3
+        )
+        # Number outputs should be 2 * 3 = 6
+        supported = len(response.choices) == 6
+    except openai.InternalServerError:
+        supported = False
+    client.server_supports_batched = supported
```
Comment (markmc) [inline src/instructlab/sdg/llmblock.py:225]: Hmm, I understand that you want to cache this ... but I don't like setting a new attribute on a class we don't own

I guess this could be removed with a move to `FlowParams`

**Review by markmc — CHANGES_REQUESTED** [review]
Cool, looks like a great direction

At least resolve the `#TODO remove sample from samples` thing

**Comment by markmc** [comment]
Also, please squash those fixup commits as per instructlab/dev-docs#110

File: src/instructlab/sdg/llmblock.py (hunk 4 above)
Comment (markmc) [reply src/instructlab/sdg/llmblock.py:54]: See `PipelineContext` in #86 now

**Comment by aakankshaduggal** [comment]
Closing in favor of #105
